In [10]:
from scipy import stats
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas_datareader import DataReader
from datetime import datetime
plt.style.use('fivethirtyeight')
import operator


In [8]:
ticker = 'ADBE'
start = datetime(2019,1,1)
end = datetime(2021,4,1)
df = DataReader(ticker,  'yahoo', start, end)

In [9]:
df

High         Low        Open       Close   Volume  \
Date                                                                  
2019-01-02  226.169998  219.000000  219.910004  224.570007  2784100   
2019-01-03  223.630005  215.149994  220.880005  215.699997  3663500   
2019-01-04  227.649994  217.479996  219.839996  226.190002  4043400   
2019-01-07  232.600006  227.289993  229.949997  229.259995  3638500   
2019-01-08  233.770004  228.330002  232.649994  232.679993  3685900   
...                ...         ...         ...         ...      ...   
2021-03-26  469.769989  449.049988  450.250000  469.089996  3614400   
2021-03-29  472.000000  462.500000  469.029999  469.320007  3141900   
2021-03-30  469.089996  461.500000  462.579987  465.459991  2436300   
2021-03-31  482.410004  468.700012  469.700012  475.369995  3155900   
2021-04-01  487.040009  480.709991  483.119995  483.339996  2664000   

             Adj Close  
Date                    
2019-01-02  224.570007  
2019-01-03  215.699997  
2019-01-04  226.190002  
2019-01-07  229.259995  
2019-01-08  232.679993  
...                ...  
2021-03-26  469.089996  
2021-03-29  469.320007  
2021-03-30  465.459991  
2021-03-31  475.369995  
2021-04-01  483.339996  

[567 rows x 6 columns]

In [25]:
a = df.Close[1:].tolist()

In [26]:
b = df.Close[:-1].tolist()

In [35]:
c = np.subtract(a,b)

In [39]:
ret = np.true_divide(c, b)

In [40]:
ret

array([-3.94977516e-02,  4.86323859e-02,  1.35726249e-02,  1.49175532e-02,
        1.18188073e-02,  9.93930964e-03, -9.25269028e-04, -1.25868468e-02,
        3.32537651e-02, -1.69171341e-03,  1.01260466e-02,  1.27250437e-02,
       -1.47872348e-02,  5.29010971e-03, -1.05246057e-02,  9.85322977e-03,
       -1.26556068e-02, -1.48025708e-02,  1.84664197e-02,  2.12222738e-02,
       -1.77549200e-03,  2.97518007e-02,  3.14052875e-04, -1.84444395e-03,
       -2.39827244e-03,  1.28477750e-02,  5.40861731e-03,  1.15328778e-02,
       -4.36157230e-03, -3.45833817e-04, -2.65242716e-03, -6.32111995e-03,
        1.12489255e-03, -1.31730465e-03,  4.42265708e-03, -1.08149396e-03,
        1.00533839e-02,  6.43133247e-03, -1.52146788e-03,  5.75241815e-03,
       -2.21582739e-02, -3.02137732e-03, -3.80764227e-03, -3.66609674e-03,
       -2.81844986e-03,  2.44169970e-02,  9.77168299e-03,  3.30156383e-03,
        1.25198483e-02, -3.95980650e-02,  2.60614352e-03,  1.03196910e-02,
       -2.61125550e-03,  

In [43]:
print(round(np.mean(ret),5))

0.00163


In [45]:
print(round(np.mean(ret),5), stats.ttest_1samp(ret,0))

0.00163 Ttest_1sampResult(statistic=1.6533388927143662, pvalue=0.09881739134445228)


The T-value is 1.65 while the P-value is 0.09. Thus, the mean is
statistically not different from zero.

## 52-week high and low trading strategy

Some investors/researchers argue that we could adopt a 52-week high and low
trading strategy by taking a long position if today's price is close to the minimum
price achieved in the past 52 weeks and taking an opposite position if today's price
is close to its 52-week high. The following Python program presents this 52-week's
range and today's position:

In [17]:
from datetime import datetime
from dateutil.relativedelta import relativedelta
ticker='ADBE'
start = datetime(2019,1,1)
end = datetime(2021,4,1)
enddate=datetime.now()
begdate=enddate-relativedelta(years=1)
df = DataReader(ticker,  'yahoo', begdate, enddate)
# x=p[-1]
# y=np.array(p.tolist())[:,-1]
# high=max(y)
# low=min(y)
# print(" Today, Price High Low, % from low ")
# print(x[0], x[-1], high, low, round((x[-1]-low)/(high-low)*100,2))

In [27]:
df

High         Low        Open       Close   Volume  \
Date                                                                  
2020-04-14  342.100006  328.470001  328.769989  340.769989  3858900   
2020-04-15  339.220001  329.649994  335.339996  332.549988  2818600   
2020-04-16  345.519989  336.040009  339.709991  342.700012  3918600   
2020-04-17  350.380005  340.500000  350.089996  344.109985  3342400   
2020-04-20  348.500000  338.750000  340.790009  344.880005  2772000   
...                ...         ...         ...         ...      ...   
2021-04-08  504.600006  498.140015  500.160004  499.839996  2026000   
2021-04-09  504.440002  492.600006  498.369995  504.040009  1886900   
2021-04-12  506.940002  498.440002  499.290009  506.029999  1584600   
2021-04-13  517.260010  506.779999  506.850006  514.859985  2195400   
2021-04-14  517.770020  509.589996  514.000000  512.732483   849293   

             Adj Close  
Date                    
2020-04-14  340.769989  
2020-04-15  332.549988  
2020-04-16  342.700012  
2020-04-17  344.109985  
2020-04-20  344.880005  
...                ...  
2021-04-08  499.839996  
2021-04-09  504.040009  
2021-04-12  506.029999  
2021-04-13  514.859985  
2021-04-14  512.732483  

[253 rows x 6 columns]

In [50]:
x = df["Adj Close"]

In [36]:
# x=p[-1]
# x

512.7324829101562

In [51]:
# len(x)

In [52]:
x

Date
2020-04-14    340.769989
2020-04-15    332.549988
2020-04-16    342.700012
2020-04-17    344.109985
2020-04-20    344.880005
                 ...    
2021-04-08    499.839996
2021-04-09    504.040009
2021-04-12    506.029999
2021-04-13    514.859985
2021-04-14    512.732483
Name: Adj Close, Length: 253, dtype: float64

In [57]:
y=np.array(df)[:,-1]

In [58]:
y

array([340.76998901, 332.54998779, 342.70001221, 344.10998535,
       344.88000488, 328.98999023, 336.76998901, 335.36999512,
       344.1000061 , 348.5       , 333.45001221, 349.17001343,
       353.64001465, 343.83999634, 349.10998535, 356.13000488,
       362.51998901, 366.77999878, 367.51000977, 371.42001343,
       365.1000061 , 358.55999756, 355.39001465, 365.29998779,
       367.97000122, 372.20001221, 383.63000488, 382.17001343,
       385.26000977, 376.63000488, 375.17001343, 379.82998657,
       386.6000061 , 389.67999268, 390.44000244, 389.92001343,
       385.79998779, 392.8999939 , 397.77999878, 397.16000366,
       406.82000732, 387.67001343, 406.54000854, 401.33999634,
       411.67001343, 413.48999023, 420.45999146, 428.01000977,
       438.64001465, 440.54998779, 431.67999268, 436.95001221,
       426.92001343, 424.20001221, 435.30999756, 439.80999756,
       442.95001221, 452.58999634, 449.35998535, 457.67999268,
       460.83999634, 466.20001221, 442.47000122, 433.77

In [59]:
# y = p[:-1].to_list()
# y

In [60]:
high=max(y)
high

533.7999877929688

In [61]:
low = min(y)
low

328.989990234375

In [62]:
print(" Today, Price High Low, % from low ")
print(x[0], x[-1], high, low, round((x[-1]-low)/(high-low)*100,2))

 Today, Price High Low, % from low 
340.7699890136719 512.7324829101562 533.7999877929688 328.989990234375 89.71


## Roll's model to estimate spread

Liquidity is defined as how quickly we can dispose of our asset without losing its
intrinsic value. Usually, we use spread to represent liquidity. However, we need
high-frequency data to estimate spread. Later in the chapter, we show how to
estimate spread directly by using high-frequency data. To measure spread indirectly
based on daily observations, Roll (1984) shows that we can estimate it based on the
serial covariance in price changes